move more sharding logic into django models

snopoke opened the pull request:
Can review per commit :fishing_pole_and_fish: 

buddy @orangejenny

snopoke committed:
use temp fsdb blobdb

snopoke committed:
add partition field to ParitionedModel

snopoke committed:
remove save and delete util methods

snopoke committed:
remove special saves for partitioned models

snopoke committed:
add partitioned_get to PartionedModel manager

snopoke committed:
remove unused imports

snopoke committed:
remove unnecessary usage of  get_db_alias_for_partitioned_doc

snopoke committed:
return a list of tuples

snopoke committed:
get forms and cases directly

snopoke committed:
add allow_relation method to db router

snopoke committed:
docs

snopoke commented on corehq/form_processor/tests/test_form_dbaccessor.py:
```diff
@@ -401,13 +401,10 @@ class FormAccessorsTestsSQL(FormAccessorsTests):
-        self.db = None
-        rmtree(self.rootdir)
-        self.rootdir = None
+        self.db.close()
```
Unrelated to this PR but I changed this while looking at a test failure

gcapalbo commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -388,9 +388,8 @@ def hard_delete_forms(domain, form_ids, delete_attachments=True):
+        for db_name, split_form_ids in split_list_by_db_partition(form_ids):
             # cascade should delete the attachments and operations
             _, deleted_models = XFormInstanceSQL.objects.using(db_name).filter(
                 domain=domain, form_id__in=form_ids
```
should this use `split_form_ids`?

snopoke replied:
yes it should thanks

gcapalbo commented on corehq/sql_db/routers.py:
```diff
@@ -22,6 +22,15 @@ def db_for_write(self, model, **hints):
+            return obj1.db == obj2.db
+        elif not obj1_partitioned and not obj2_partitioned:
+            return True
+
```
should this explicitly return False or is returning None ok?

snopoke replied:
In our case it has the same effect since there are no other routers but better to be explicit

gcapalbo reviewed (approved):
a lot of great stuff here!

snopoke committed:
use correct list

snopoke committed:
fix test for unsharded DB

snopoke committed:
set form ID not form object

During the deprecation workflow a form's ID is changed so
when we re-attach the form to the related models
django does a check to see if that's allowed and because
the related models still refer to the old form ID there
is an edge case where the deprecated form is being saved to a
different DB than where the attachment is pointing to

millerdev commented:
Just looking at the title of this PR makes me wonder... It seems like we're doing something wrong if we can't get our proxy db to perform at the level we need, especially after seeing the citus demo and noting that it has basically the came architecture as plproxy (central proxy node), which is apparently not a problem for them.

Moving sharding logic into python seems wrong because python is quite slow compared to postgres, so I will not be surprised if we don't encounter more/worse performance issues due to this move down the road.

snopoke commented:
@millerdev the bottleneck is number of connections to PG the way to get round this is to either have a dedicated node for pl_proxy or to have multiple pl_proxy nodes that you load balance across.

The approach (moving sharding logic to python) will keep the infrastructure setup simpler at the cost of adding some complexity to python code (though that complexity is technically already there because of what we have setup with the SQL functions). 

There is another reason for doing it in python which is that it gives us more control over the sharding which we will need when we split the shard databases. 

Happy to discuss this more if you think this is going to cause problems. Re performance what do you think is going be worse? The only thing I can think of is perhaps when we query multiple databases that pl_proxy might do that in parallel.

millerdev commented:
@snopoke 
> the bottleneck is number of connections to PG

I thought we used pgbouncer to get around that problem. That's even what the citusdb guy suggested when we mentioned the problem of the proxy being a bottleneck.

> gives us more control over the sharding which we will need when we split the shard databases

Not sure what you mean by this. Are you saying we're going to have multiple different sharded databases? Why would we do that? In terms of adding shards and moving them to different machines, if we do that correctly it should be transparent to the clients (python code accessing the database).

> Re performance what do you think is going be worse?

Adding the sharding logic to python means we're moving another layer of our database into our application code, which means it is more likely to become tightly coupled with the application implementation and can be harder to analyze performance bottlenecks. Python is slow, and when we're optimizing for performance we should either be offloading the processing to fast things like postgres or or maybe a C-implementation of some time-critical logical component, or doing it in parallel with more python processes (although this can get heavy and expensive in terms of hardware).

snopoke commented:
@millerdev 

We do use pgbouncer. We haven't really had any issues with this except on NIC when there are resource conflicts on the hosts that cause the DB to suffer and queries to take longer. This pushes the number of concurrent requests up and hence the number of connections to the DB.

As an example of how this works in practice, on NIC we have 1024 logical shards split across 10 physical databases (5 on each of 2 DB VMs). One of those VMs also has the pl_proxy DB on it. All requests to those 10 databases go through pl_proxy so if we have 25 connections to each of the 10 DBs then the PG instance with pl_proxy has 25*5 + 250 PG connections = 375 connections. Obviously there are ways around this. Firstly moving the pl_proxy node to a separate VM would make a big difference etc.

> Not sure what you mean by this. Are you saying we're going to have multiple different sharded databases? Why would we do that? In terms of adding shards and moving them to different machines, if we do that correctly it should be transparent to the clients (python code accessing the database).

We already have multiple physical shards. In terms of moving shards (e.g. splitting a physical shard) yes, it would be transparent to the client but during the process of actually moving the data the client wouldn't be able to connect at all. If we have the sharding logic in python we have more flexibility in terms of telling the client where to look for the data.

> Adding the sharding logic to python means we're moving another layer of our database into our application code, which means it is more likely to become tightly coupled with the application implementation and can be harder to analyze performance bottlenecks. 

Sure, I agree with that generally though the sharding logic is pretty well isolated and easy to test.

> Python is slow, and when we're optimizing for performance we should either be offloading the processing to fast things like postgres or or maybe a C-implementation of some time-critical logical component, or doing it in parallel with more python processes (although this can get heavy and expensive in terms of hardware).

Considering the only thing we're moving to python is the query routing I don't expect there is going to much loss in performance. The cost of running the query is going to far outweigh the cost of routing the query to the appropriate DB. The function that's doing the 'hashing' as part of the shard logic is already in C.

millerdev commented:
> We haven't really had any issues with this except on NIC when there are resource conflicts on the hosts that cause the DB to suffer and queries to take longer. This pushes the number of concurrent requests up and hence the number of connections to the DB.

Considering the hardware/virtualization situation there (i.e., way beyond reasonably limited resources) and that this issue may be entirely a symptom of that, it makes me sad that we would be making such a profound architectural change to work around that one very unfortunate situation.

With regard to the difficulty of scaling the number of servers in our pl_proxy cluster, are you saying there's no  way to do move (virtual) shards without downtime? I'm not talking about changing the number of (virtual) shards, which I understand is not something we ever want to do. If that's true that seems like a pretty big limitation of pl_proxy, and a good argument to switch to something like citusdb. The most impressive part of the citus demo was how scaling was completely transparent to clients, and it did not require any downtime.

snopoke commented:
Pl_proxy has no tools or features to allow moving shard around. It would be
a completely manual process.

millerdev commented on corehq/sql_db/models.py:
```diff
@@ -57,17 +93,35 @@ class PartitionedModel(models.Model):
 
     def save(self, *args, **kwargs):
-        if not kwargs.get('using'):
-            raise_access_restricted()
```
👍 nice to see this go

millerdev commented on corehq/sql_db/models.py:
```diff
@@ -57,17 +93,35 @@ class PartitionedModel(models.Model):
     def delete(self, *args, **kwargs):
-        if not kwargs.get('using'):
-            raise_access_restricted()
+        if kwargs.get('using'):
```
same as above

millerdev reviewed (commented):
Finally got around to reviewing the actual code, and it looks like there was a fair amount of cleanup that went into this effort as well, which is very nice.

I'm still not sold on moving partition logic into Python being a good idea, mostly because I think our motivation (proxy seems like bottleneck on badly configured cluster) is not solid.

snopoke commented on corehq/sql_db/models.py:
```diff
@@ -57,17 +93,35 @@ class PartitionedModel(models.Model):
     def save(self, *args, **kwargs):
-        if not kwargs.get('using'):
-            raise_access_restricted()
+        if kwargs.get('using'):
```
https://github.com/dimagi/commcare-hq/pull/17376
